prov/efa/src/efa_msg.c

```diff
@@ -0,0 +1,408 @@
+
+	return efa_post_recv(base_ep, &msg, 0);
+}
+
+static inline ssize_t efa_post_send_validate(struct efa_base_ep *base_ep, const struct fi_msg *msg,
+				      struct efa_conn *conn, uint64_t flags, size_t *len)
+{
+	if (OFI_UNLIKELY(msg->iov_count > base_ep->info->tx_attr->iov_limit)) {
+		EFA_WARN(FI_LOG_EP_DATA, "requested sge[%zu] is greater than max supported[%zu]!\n",
+			 msg->iov_count, base_ep->info->tx_attr->iov_limit);
+		return -FI_EINVAL;
+	}
+
+	if (OFI_UNLIKELY(msg->msg_iov[0].iov_len <
+			 base_ep->info->ep_attr->msg_prefix_size)) {
+		EFA_WARN(FI_LOG_EP_DATA, "prefix not present on first iov, iov_len[%zu]\n",
+			 msg->msg_iov[0].iov_len);
+		return -EINVAL;
+	}
+
+	*len = ofi_total_iov_len(msg->msg_iov, msg->iov_count) - base_ep->info->ep_attr->msg_prefix_size;
+	if (OFI_UNLIKELY(*len > base_ep->info->ep_attr->max_msg_size)) {
+		EFA_WARN(FI_LOG_EP_DATA, "requested size[%zu] is greater than max[%zu]!\n",
+			 *len, base_ep->info->ep_attr->max_msg_size);
+		return -FI_EINVAL;
+	}
+
+	return 0;
+}
+
+static inline bool efa_msg_has_hmem_mr(const struct fi_msg *msg)
+{
+	/* the device only support send up 2 iov, so iov_count cannot be > 2 */
+	assert(msg->iov_count == 1 || msg->iov_count == 2);
+	/* first iov is always on host memory, because it must contain packet header */
+	assert(!efa_mr_is_hmem(msg->desc[0]));
+	return (msg->iov_count == 2) && efa_mr_is_hmem(msg->desc[1]);
+}
+
+static inline ssize_t efa_post_send(struct efa_base_ep *base_ep, const struct fi_msg *msg, uint64_t flags)
+{
+	struct efa_qp *qp = base_ep->qp;
+	struct efa_conn *conn;
+	struct ibv_sge sg_list[2];  /* efa device support up to 2 iov */
+	struct ibv_data_buf inline_data_list[2];
+	size_t len, i, iov_offset = 0;
+	int ret, iov_idx = 0;
+
+	dump_msg(msg, "send");
+
+	conn = efa_av_addr_to_conn(base_ep->av, msg->addr);
+	assert(conn && conn->ep_addr);
+
+	ret = efa_post_send_validate(base_ep, msg, conn, flags, &len);
+	if (OFI_UNLIKELY(ret))
+		return ret;
+
+	if (!base_ep->is_wr_started) {
+		ibv_wr_start(qp->ibv_qp_ex);
+		base_ep->is_wr_started = true;
+	}
+
+	qp->ibv_qp_ex->wr_id = (uintptr_t)msg->context;
+	if (flags & FI_REMOTE_CQ_DATA) {
+		ibv_wr_send_imm(qp->ibv_qp_ex, msg->data);
+	} else {
+		ibv_wr_send(qp->ibv_qp_ex);
+	}
+
+	ret = ofi_iov_locate((struct iovec *) msg->msg_iov, msg->iov_count,
+			     base_ep->info->ep_attr->msg_prefix_size, &iov_idx,
+			     &iov_offset);
+	if (OFI_UNLIKELY(ret)) {
+		EFA_WARN(FI_LOG_CQ, "ofi_iov_locate failed! err: %d\n", ret);
+		return ret;
+	}
+	assert(iov_idx < msg->iov_count);
+	assert(iov_offset < msg->msg_iov[iov_idx].iov_len);
+
+	if (len <= base_ep->domain->device->efa_attr.inline_buf_size &&
+	    !efa_msg_has_hmem_mr(msg)) {
+		for (i = 0; i < msg->iov_count; i++) {
+			inline_data_list[i].addr = msg->msg_iov[i].iov_base;
+			inline_data_list[i].length = msg->msg_iov[i].iov_len;
+		}
+		inline_data_list[iov_idx].addr = (char *) msg->msg_iov[iov_idx].iov_base + iov_offset;
```

Comment:
Is this correct? Should the loop start from `iov_index`? what about the case where iov_count is 2 but the first iov is all prefix? Then should you drop the first iov entirely?